.evergreen/socks5srv.py

```diff
@@ -0,0 +1,237 @@
+    port_re = r"(?P<{0}_port>\d+)"
+
+    src_re = host_re.format('src') + ':' + port_re.format('src')
+    dst_re = host_re.format('dst') + ':' + port_re.format('dst')
+    full_re = '^' + src_re + ' to ' + dst_re + '$'
+
+    match = re.match(full_re, string)
+    if match is None:
+      raise Exception("Mapping {} does not match format '{{host}}:{{port}} to {{host}}:{{port}}'".format(string))
+
+    src = ((match.group('src_ipv6') or match.group('src_host')).encode('utf8'), int(match.group('src_port')))
+    dst = ((match.group('dst_ipv6') or match.group('dst_host')).encode('utf8'), int(match.group('dst_port')))
+    return (src, dst)
+
+  def add_dns_remappings(self):
+    """Add mappings for the IP addresses corresponding to hostnames
+
+    For example, if there is a mapping (localhost, 1000) to (localhost, 2000),
+    then this also adds (127.0.0.1, 1000) to (localhost, 2000)."""
+
+    for src, dst in self.mappings:
+      host, port = src
+      try:
+        addrs = socket.getaddrinfo(host, port, socket.AF_UNSPEC, socket.SOCK_STREAM)
+      except socket.gaierror:
+        continue
+
+      existing_src_entries = [src for src, dst in self.mappings]
+      for af, socktype, proto, canonname, sa in addrs:
+        if af == socket.AF_INET and sa not in existing_src_entries:
+          self.mappings.append((sa, dst))
+        elif af == socket.AF_INET6 and sa[:2] not in existing_src_entries:
+          self.mappings.append((sa[:2], dst))
+
+  def remap(self, hostport):
+    """Re-map a (host, port) tuple to a new (host, port) tuple if that was requested"""
+
+    for src, dst in self.mappings:
+      if hostport == src:
+        return dst
+    return hostport
+
+class Socks5Server(socketserver.ThreadingTCPServer):
+  """A simple Socks5 proxy server"""
+
+  def __init__(self, server_address, RequestHandlerClass, args):
+    socketserver.ThreadingTCPServer.__init__(self,
+                                             server_address,
+                                             RequestHandlerClass)
+    self.args = args
+    self.address_remapper = AddressRemapper(args.map)
+
+class Socks5Handler(socketserver.BaseRequestHandler):
+  """Request handler for Socks5 connections"""
+
+  def finish(self):
+    """Called after handle(), always just closes the connection"""
+
+    self.request.close()
+
+  def read_exact(self, n):
+    """Read n bytes from a socket
+
+    In Socks5, strings are prefixed with a single byte containing
+    their length. This method reads a bytes string containing n bytes
+    (where n can be a number or a bytes object containing that
+    single byte).
+
+    If reading from the client ends prematurely, this returns None.
+    """
+
+    if type(n) is bytes:
+      if len(n) == 0:
+        return None
+      assert len(n) == 1
+      n = n[0]
+    result = b''
+    while len(result) < n:
+      buf = self.request.recv(n - len(result))
+      if buf == b'':
+        return None
+      result += buf
```

Comment:
So building the result buffer like this `+=` can yield o(n^2) performance. See PYTHON-1513. I'm not sure if this method is only used for reading small pieces of data but a more efficient way would be:
```python
    buf = bytearray(n)
    mv = memoryview(buf)
    bytes_read = 0
    while bytes_read < n:
        try:
            chunk_length = self.request.recv_into(mv[bytes_read:])
        except OSError as exc:
            return None
        if chunk_length == 0:
            return None

        bytes_read += chunk_length
    return bytes(buf)
```

Reply:
Applied your suggestion, but yeah, I wouldn’t have been too worried about that here – the buffer size is limited by the procotol definition to 255 bytes, and the actual usage in the tests doesn’t even come close to that..